// @copilotkit/a2ui-renderer/dist/themed-surface.js
/*
 Copyright 2025 Google LLC

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

      https://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
import { SignalWatcher } from "@lit-labs/signals";
import { provide } from "@lit/context";
import { html, LitElement } from "lit";
import { customElement, property } from "lit/decorators.js";
import * as UI from "@a2ui/lit/ui";
import { globalStyles } from "./styles/global.js";
let ThemedA2UISurface = (() => {
    let _classDecorators = [customElement("themed-a2ui-surface")];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = SignalWatcher(LitElement);
    let _theme_decorators;
    let _theme_initializers = [];
    let _theme_extraInitializers = [];
    let _surfaceId_decorators;
    let _surfaceId_initializers = [];
    let _surfaceId_extraInitializers = [];
    let _surface_decorators;
    let _surface_initializers = [];
    let _surface_extraInitializers = [];
    let _processor_decorators;
    let _processor_initializers = [];
    let _processor_extraInitializers = [];
    let _onAction_decorators;
    let _onAction_initializers = [];
    let _onAction_extraInitializers = [];
    var ThemedA2UISurface = class extends _classSuper {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _theme_decorators = [provide({ context: UI.Context.themeContext }), property({ attribute: false })];
            _surfaceId_decorators = [property({ attribute: false })];
            _surface_decorators = [property({ attribute: false })];
            _processor_decorators = [property({ attribute: false })];
            _onAction_decorators = [property({ attribute: false })];
            __esDecorate(this, null, _theme_decorators, { kind: "accessor", name: "theme", static: false, private: false, access: { has: obj => "theme" in obj, get: obj => obj.theme, set: (obj, value) => { obj.theme = value; } }, metadata: _metadata }, _theme_initializers, _theme_extraInitializers);
            __esDecorate(this, null, _surfaceId_decorators, { kind: "accessor", name: "surfaceId", static: false, private: false, access: { has: obj => "surfaceId" in obj, get: obj => obj.surfaceId, set: (obj, value) => { obj.surfaceId = value; } }, metadata: _metadata }, _surfaceId_initializers, _surfaceId_extraInitializers);
            __esDecorate(this, null, _surface_decorators, { kind: "accessor", name: "surface", static: false, private: false, access: { has: obj => "surface" in obj, get: obj => obj.surface, set: (obj, value) => { obj.surface = value; } }, metadata: _metadata }, _surface_initializers, _surface_extraInitializers);
            __esDecorate(this, null, _processor_decorators, { kind: "accessor", name: "processor", static: false, private: false, access: { has: obj => "processor" in obj, get: obj => obj.processor, set: (obj, value) => { obj.processor = value; } }, metadata: _metadata }, _processor_initializers, _processor_extraInitializers);
            __esDecorate(this, null, _onAction_decorators, { kind: "accessor", name: "onAction", static: false, private: false, access: { has: obj => "onAction" in obj, get: obj => obj.onAction, set: (obj, value) => { obj.onAction = value; } }, metadata: _metadata }, _onAction_initializers, _onAction_extraInitializers);
            __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            ThemedA2UISurface = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            __runInitializers(_classThis, _classExtraInitializers);
        }
        #theme_accessor_storage = __runInitializers(this, _theme_initializers, void 0);
        get theme() { return this.#theme_accessor_storage; }
        set theme(value) { this.#theme_accessor_storage = value; }
        #surfaceId_accessor_storage = (__runInitializers(this, _theme_extraInitializers), __runInitializers(this, _surfaceId_initializers, null));
        get surfaceId() { return this.#surfaceId_accessor_storage; }
        set surfaceId(value) { this.#surfaceId_accessor_storage = value; }
        #surface_accessor_storage = (__runInitializers(this, _surfaceId_extraInitializers), __runInitializers(this, _surface_initializers, null));
        get surface() { return this.#surface_accessor_storage; }
        set surface(value) { this.#surface_accessor_storage = value; }
        #processor_accessor_storage = (__runInitializers(this, _surface_extraInitializers), __runInitializers(this, _processor_initializers, null));
        get processor() { return this.#processor_accessor_storage; }
        set processor(value) { this.#processor_accessor_storage = value; }
        #onAction_accessor_storage = (__runInitializers(this, _processor_extraInitializers), __runInitializers(this, _onAction_initializers, null));
        get onAction() { return this.#onAction_accessor_storage; }
        set onAction(value) { this.#onAction_accessor_storage = value; }
        #handleAction = (__runInitializers(this, _onAction_extraInitializers), (event) => {
            this.onAction?.(event, {
                surfaceId: this.surfaceId,
                surface: this.surface,
                processor: this.processor,
            });
        });
        render() {
            return html `<style>
        ${globalStyles}
      </style>
      <a2ui-surface
        @a2uiaction=${this.#handleAction}
        .surfaceId=${this.surfaceId}
        .surface=${this.surface}
        .processor=${this.processor}
      ></a2ui-surface>`;
        }
    };
    return ThemedA2UISurface = _classThis;
})();
export { ThemedA2UISurface };
